### PR TITLE
D500 recipe

### PR DESCRIPTION
**Pull request by sashkboos**
Deep500 recipe added for `Uncertainty Quantification` task.

### CODE REVIEW

File: Uncertainty_Quantification/Deep500_recipe/loader.py [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+        data_x = reduce_sample_x(data_x, self.args, self.means, self.stddevs)
+
+        if not self.infer and self.step == "train":
+
+            data_y = np.load(self.datalist_y[idx])
+            data_y = reduce_sample_y(data_y, self.args)
+
+            for aug in self.args.augmentation:  # Apply transformations if chosen
+                data_x, data_y = TRANSFORMATION_DICTIONARY[aug](data_x, data_y,
+                                                                self.args)
```
[the remark this answers is not in the document]
Comment (tbennun) [reply Uncertainty_Quantification/Deep500_recipe/loader.py:83]: quite important if we want to e.g., cache samples

File: Uncertainty_Quantification/Deep500_recipe/loader.py [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+            if not self.infer and (self.step == "val" or self.step == "test"):
+                data_y = np.load(self.datalist_y[idx])
+                data_y = reduce_sample_y(data_y, self.args)
+
+                return data_x.copy(), data_y.copy()
+
+            return data_x.copy()
+
+
+class WeatherShuffleSampler(ShuffleSampler):
```
Comment (tbennun) [inline Uncertainty_Quantification/Deep500_recipe/loader.py:172]: why is this not the normal shuffle sampler?

File: Uncertainty_Quantification/Deep500_recipe/loader.py [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+
+    def __len__(self):
+        return len(self.dataset) // self.batch_size
+
+    def reset(self):
+        super().reset()
+        self.batch_idx = 0
+
+
+class callable_weather_dataset:
```
Comment (tbennun) [inline Uncertainty_Quantification/Deep500_recipe/loader.py:214]: name convention

File: Uncertainty_Quantification/Deep500_recipe/losses.py [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+class MSSSIMLoss(torch.nn.Module):
+    def __init__(self, dim=2):
+        super(MSSSIMLoss, self).__init__()
+        if dim == 2:
+            self.msssim = MS_SSIM(data_range=4.3, channel=1)
+        else:
+            self.msssim = MS_SSIM(
+                data_range=4.3,
+                channel=2  # 60 for mean
+            )  # after standardization ~0+-60 through analysis, channel = 2 for 2 pressure levels
```
Comment (tbennun) [inline Uncertainty_Quantification/Deep500_recipe/losses.py:15]: 2 for 2 pressure levels, what does that mean?

File: Uncertainty_Quantification/Deep500_recipe/recipe_UQ.py [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+    "dataset": loader.callable_weather_dataset(args),
+    "epochs": args.epochs,
+}
+
+# Mutable Components
+MUTABLE = {
+    "batch_size":
+    args.batch_size,
+    "executor":
+    executor(),
```
Comment (tbennun) [inline Uncertainty_Quantification/Deep500_recipe/recipe_UQ.py:58]: styling question - why are the values in a different line than the keys? I'd put them on the same line

File: Uncertainty_Quantification/Deep500_recipe/recipe_UQ.py [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+
+import torch
+from events import *
+from models import *
+from utils import *
+from losses import *
+import loader
+from scheduler import CyclicLRScheduler
+
+args = args_parser()
```
Comment (tbennun) [inline Uncertainty_Quantification/Deep500_recipe/recipe_UQ.py:19]: that's weird, why would this be here? the recipe should be parameterized with everything. I'd replace with a fixed `dataclass` (as said above).

**Review by tbennun — CHANGES_REQUESTED** [review]

File: Uncertainty_Quantification/Deep500_recipe/loader.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (sashkboos) [reply Uncertainty_Quantification/Deep500_recipe/loader.py]: Fixed!

File: Uncertainty_Quantification/Deep500_recipe/loader.py [hunk 7]
```diff
@@ -0,0 +1,253 @@
...
+        data_x = np.load(self.datalist_x[idx])
+        data_x = reduce_sample_x(data_x, self.args, self.means, self.stddevs)
+
+        if not self.infer and self.step == "train":
+            data_y = np.load(self.datalist_y[idx])
+            data_y = reduce_sample_y(data_y, self.args)
+
+            for aug in self.args.augmentation:  # Apply transformations if chosen
+                data_x, data_y = TRANSFORMATION_DICTIONARY[aug](data_x, data_y,
+                                                                self.args)
```
[the remark this answers is not in the document]
Comment (sashkboos) [reply Uncertainty_Quantification/Deep500_recipe/loader.py]: Fixed

File: Uncertainty_Quantification/Deep500_recipe/README.md [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+
+[Deep500](https://www.deep500.org)  is A modular benchmarking infrastructure for 
+high-performance deep learning — from the single operator 
+to distributed training. In this directory, we  reproduce the result of
+*uncertainty quantification* task using deep500 recipe.
+
+
+## Usage
+
+To run the code, you need to install Deep500 using the following line:
```
[the remark this answers is not in the document]
Comment (sashkboos) [reply Uncertainty_Quantification/Deep500_recipe/README.md]: Fixed

File: Uncertainty_Quantification/Deep500_recipe/losses.py (hunk 4 above)
Comment (sashkboos) [reply Uncertainty_Quantification/Deep500_recipe/losses.py:15]: This is just a copy from this comment:

https://github.com/spcl/deep-weather/blob/master/Uncertainty_Quantification/Pytorch/models.py#L28

File: Uncertainty_Quantification/Deep500_recipe/loader.py (hunk 2 above)
Comment (sashkboos) [reply Uncertainty_Quantification/Deep500_recipe/loader.py:173]: The sampler wants to access multiple indices here:

https://github.com/deep500/deep500/blob/34e93a46dea17513ac705bb4392b3514fa9d87c6/deep500/lv2/sampler.py#L144-L145

And this runs into an error. However, we can do it (by changing the sampler and/or dataset we defined).

File: Uncertainty_Quantification/Deep500_recipe/loader.py [hunk 9]
```diff
@@ -0,0 +1,253 @@
...
+TRANSFORMATION_DICTIONARY = {
+    "RandomCrop": random_crop,
+    "RandomHorizontalFlip": horizontal_flip,
+    "RandomVerticalFlip": vertical_flip,
+    "Transpose": transpose,
+}
+
+
+class WeatherDataset(Dataset):
+    def __init__(self, args, step="train", infer=False, year_dict=None):
```
[the remark this answers is not in the document]
Comment (sashkboos) [reply Uncertainty_Quantification/Deep500_recipe/loader.py]: We have added the UQDataClass.

**Review by tbennun — APPROVED** [review]
